display_list/dl_builder.cc

```diff
@@ -44,10 +65,15 @@ void* DisplayListBuilder::Push(size_t pod, Args&&... args) {
   size_t size = SkAlignPtr(sizeof(T) + pod);
   FML_CHECK(size < (1 << 24));
   if (used_ + size > allocated_) {
-    static_assert(is_power_of_two(DL_BUILDER_PAGE),
+    static_assert(is_power_of_two(kDLBuilderMaxGrowth),
                   "This math needs updating for non-pow2.");
-    // Next greater multiple of DL_BUILDER_PAGE.
-    allocated_ = (used_ + size + DL_BUILDER_PAGE) & ~(DL_BUILDER_PAGE - 1);
+    // Next power of two, up to a limit of DL_BUILDER_MAX_GROWTH. This
+    // assumes that the requested size to allocate is always less than
+    // kDLBuilderMaxGrowth, which given is over a MB is a reasonable
+    // assumption.
+    uint64_t next_size = NextPowerOfTwo(used_ + size);
+    uint64_t growth_in_bytes = (next_size - allocated_);
+    allocated_ += std::min(growth_in_bytes, kDLBuilderMaxGrowth);
```

Comment:
If we need more than MaxGrowth then we fail here. You can max the "unused" part of the growth to MaxGrowth, but you can't make it less than the amount we need total (used_ + size). Better math is needed here...

Reply:
okay, I think I was trying to be too clever. I've changed this code so that we always double, and we first increment by whatever is bigger - required size or page size, where page size is about 16K